docs/compilers/CSharp/Runtime Async Design.md

```diff
@@ -0,0 +1,456 @@
+
+```cs
+namespace System.Runtime.CompilerServices;
+
+// These methods are used to await things that cannot use runtime async signature form
+// TODO: Clarify which of these should be preferred? Should we always emit the `Unsafe` version when awaiting something that implements `ICriticalNotifyCompletion`?
+namespace System.Runtime.CompilerServices;
+
+public static class RuntimeHelpers
+{
+    [RuntimeAsyncMethod]
+    public static Task AwaitAwaiterFromRuntimeAsync<TAwaiter>(TAwaiter awaiter) where TAwaiter : INotifyCompletion;
+    [RuntimeAsyncMethod]
+    public static Task UnsafeAwaitAwaiterFromRuntimeAsync<TAwaiter>(TAwaiter awaiter) where TAwaiter : ICriticalNotifyCompletion;
+}
+```
+
+Additionally, we use the following helper attributes to indicate information to the runtime. If these attributes are not present in the reference assemblies, we will generate them; the runtime matches by full
+name, not by type identity, so we do not need to care about using the "canonical" versions.
+
+```cs
+namespace System.Runtime.CompilerServices;
+
+// Used to tell the runtime to generate the async state machinery for this method
+[AttributeUsage(AttributeTargets.Method)]
+public class RuntimeAsyncMethodAttribute() : Attribute();
+
+// Used to mark locals that should be hoisted to the generated async closure. Note that the runtime does not guarantee that all locals marked with this attribute will be hoisted; if it can prove that it
+// doesn't need to hoist a variable, it may avoid doing so.
+public class HoistedLocal();
+```
+
+For experimentation purposes, we recognize an attribute that can be used to force the compiler to generate the runtime async code, or to force the compiler to generate a full state machine. This attribute is not
+defined in the BCL, and exists as an escape hatch for experimentation. It may be removed when the feature ships in stable.
+
+```cs
+namespace System.Runtime.CompilerServices;
+
+[AttributeUsage(AttributeTargets.Method)]
+public class RuntimeAsyncMethodGenerationAttribute(bool runtimeAsync) : Attribute();
+```
+
+## Transformation strategy
+
+As mentioned previously, we try to expose as little of this to initial binding as possible. The one major exception to this is our handling of the `RuntimeAsyncMethodAttribute`; we do not let this be applied to
+user code, and will issue an error if a user tries to do this by hand.
+
+Compiler generated async state machines and runtime generated async share some of the same building blocks. Both need to have `await`s with in `catch` and `finally` blocks rewritten to pend the exceptions,
+perform the `await` outside of the `catch`/`finally` region, and then have the exceptions restored as necessary.
+
+TODO: Go over `IAsyncEnumerable` and confirm that the initial rewrite to a `Task`-based method produces code that can then be implemented with runtime async, rather than a full compiler state machine.
```

Comment:
FWIW, I know this is not blocking early work, but for peace of mind I took a stab at this to show how it could work: [https://gist.github.com/jcouv/aa5359db1b15522013cf19ef52dcc18f](https://gist.github.com/jcouv/aa5359db1b15522013cf19ef52dcc18f)